wgengine/userspace.go

```diff
@@ -145,26 +145,31 @@ type EngineConfig struct {
 	// which disables such features as DNS configuration and unrestricted ICMP Echo responses.
 	Fake bool
 
-	// FakeImpl, if non-nil, specifies which type of fake implementation to
-	// use. Two values are typical: nil, for a basic ping-only fake
-	// implementation, and netstack.Impl, which brings in gvisor's netstack
-	// to the binary. The desire to keep that out of some binaries is why
-	// this func exists, so wgengine need not depend on gvisor.
-	FakeImpl FakeImplFunc
+	// FakeImplFactory, if non-nil, creates a FakeImpl to use as a fake
+	// implementation. See docs for FakeImpl below for more details.
+	FakeImplFactory FakeImplFactory
 }
 
-// FakeImplFunc is the type used by EngineConfig.FakeImpl. See docs there.
-type FakeImplFunc func(logger.Logf, *tstun.TUN, Engine, *magicsock.Conn) error
+// FakeImpl specifies which type of fake implementation to use. Two values
+// are typical: nil, for a basic ping-only fake implementation, and netstack.Impl,
+// which brings in gvisor's netstack to the binary. The desire to keep that out
+// of some binaries is why this func exists, so wgengine need not depend on gvisor.
+type FakeImpl interface {
+	Start() error
+}
+
+// FakeImplFactory is the type of a function used to create FakeImpls.
+type FakeImplFactory func(logf logger.Logf, tundev *tstun.TUN, e Engine, mc *magicsock.Conn) (FakeImpl, error)
```

Comment:
you can remove all the parameter names here. The types are obvious enough that the names (logf, tundev, e, mc) don't add any disambiguation/clarity.